File: src/main/java/med/nano/NanoswarmValidator.java
package med.nano;

import java.util.List;
import java.util.Map;

public final class NanoswarmValidator {

    public enum ComplianceLevel {
        RESEARCH_GRADE,
        CLINICAL_TRIAL,
        SURGICAL_GRADE,
        IMPLANTABLE_ISO14708
    }

    public static final class NanoswarmNode {
        public String nodeId;
        public double opsThresholdTops;
        public double[] topologyMatrix;
        public ComplianceLevel complianceLevel;
        public String aiFirmwareVersion;
        public float tissueTempMaxC;
        public float powerDensityMaxMwCm2;
        public float emExposureMaxVM;
        public float[] biocompatibilityMatrix;
        public String firmwareSha256Hex;
    }

    public static final class ValidationResult {
        public boolean validationStatus;
        public Map<String, Object> systemMetrics;
    }

    private static float minBiocompatibility(float[] m) {
        float min = 1.0f;
        for (float v : m) {
            if (v < min) {
                min = v;
            }
        }
        return min;
    }

    private static double zeroFraction(double[] m) {
        if (m.length == 0) {
            return 1.0;
        }
        int zeros = 0;
        for (double v : m) {
            if (Math.abs(v) < 1e-9) {
                zeros++;
            }
        }
        return (double) zeros / (double) m.length;
    }

    public static ValidationResult validate(NanoswarmNode node) {
        boolean computeOk = node.opsThresholdTops >= 12.0;
        int nonZero = 0;
        for (double v : node.topologyMatrix) {
            if (Math.abs(v) >= 1e-9) {
                nonZero++;
            }
        }
        boolean topoOk = nonZero >= 8;

        boolean complianceOk = node.complianceLevel == ComplianceLevel.CLINICAL_TRIAL
                || node.complianceLevel == ComplianceLevel.SURGICAL_GRADE
                || node.complianceLevel == ComplianceLevel.IMPLANTABLE_ISO14708;

        boolean fwOk = node.firmwareSha256Hex != null
                && node.firmwareSha256Hex.length() == 64
                && node.firmwareSha256Hex.chars().allMatch(c ->
                    (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F'));

        float bioMin = minBiocompatibility(node.biocompatibilityMatrix);
        boolean bioOk = bioMin >= 0.90f;

        boolean safeTemp = node.tissueTempMaxC <= 39.0f;
        boolean safePower = node.powerDensityMaxMwCm2 <= 50.0f;
        boolean safeEm = node.emExposureMaxVM <= 61.0f;

        boolean allOk = computeOk && topoOk && complianceOk && fwOk && bioOk && safeTemp && safePower && safeEm;

        ValidationResult result = new ValidationResult();
        result.validationStatus = allOk;
        result.systemMetrics = Map.of(
                "node_id", node.nodeId,
                "ops_threshold_tops", node.opsThresholdTops,
                "topology_sparsity", zeroFraction(node.topologyMatrix),
                "biocompatibility_min", bioMin,
                "tissue_temp_max_c", node.tissueTempMaxC,
                "power_density_max_mw_cm2", node.powerDensityMaxMwCm2,
                "em_exposure_max_v_m", node.emExposureMaxVM,
                "compliance_level", node.complianceLevel.name(),
                "ai_firmware_version", node.aiFirmwareVersion
        );
        return result;
    }
}

File: src/main/kotlin/med/nano/NanoswarmValidator.kt
package med.nano

data class NanoswarmNode(
    val nodeId: String,
    val opsThresholdTops: Double,
    val topologyMatrix: DoubleArray,
    val complianceLevel: ComplianceLevel,
    val aiFirmwareVersion: String,
    val tissueTempMaxC: Float,
    val powerDensityMaxMwCm2: Float,
    val emExposureMaxVM: Float,
    val biocompatibilityMatrix: FloatArray,
    val firmwareSha256Hex: String,
)

enum class ComplianceLevel {
    RESEARCH_GRADE,
    CLINICAL_TRIAL,
    SURGICAL_GRADE,
    IMPLANTABLE_ISO14708,
}

data class ValidationResult(
    val validationStatus: Boolean,
    val systemMetrics: Map<String, Any>,
)

object NanoswarmValidator {

    private fun minBiocompatibility(m: FloatArray): Float =
        m.fold(1.0f) { acc, v -> if (v < acc) v else acc }

    private fun zeroFraction(m: DoubleArray): Double {
        if (m.isEmpty()) return 1.0
        val zeros = m.count { kotlin.math.abs(it) < 1e-9 }
        return zeros.toDouble() / m.size.toDouble()
    }

    private fun isHex64(s: String): Boolean =
        s.length == 64 && s.all { it.isDigit() || (it.lowercaseChar() in 'a'..'f') }

    fun validate(node: NanoswarmNode): ValidationResult {
        val computeOk = node.opsThresholdTops >= 12.0
        val nonZero = node.topologyMatrix.count { kotlin.math.abs(it) >= 1e-9 }
        val topoOk = nonZero >= 8

        val complianceOk = when (node.complianceLevel) {
            ComplianceLevel.CLINICAL_TRIAL,
            ComplianceLevel.SURGICAL_GRADE,
            ComplianceLevel.IMPLANTABLE_ISO14708 -> true
            else -> false
        }

        val fwOk = isHex64(node.firmwareSha256Hex)
        val bioMin = minBiocompatibility(node.biocompatibilityMatrix)
        val bioOk = bioMin >= 0.90f

        val safeTemp = node.tissueTempMaxC <= 39.0f
        val safePower = node.powerDensityMaxMwCm2 <= 50.0f
        val safeEm = node.emExposureMaxVM <= 61.0f

        val allOk = computeOk && topoOk && complianceOk && fwOk && bioOk && safeTemp && safePower && safeEm

        val metrics = mapOf<String, Any>(
            "node_id" to node.nodeId,
            "ops_threshold_tops" to node.opsThresholdTops,
            "topology_sparsity" to zeroFraction(node.topologyMatrix),
            "biocompatibility_min" to bioMin,
            "tissue_temp_max_c" to node.tissueTempMaxC,
            "power_density_max_mw_cm2" to node.powerDensityMaxMwCm2,
            "em_exposure_max_v_m" to node.emExposureMaxVM,
            "compliance_level" to node.complianceLevel.name,
            "ai_firmware_version" to node.aiFirmwareVersion,
        )

        return ValidationResult(
            validationStatus = allOk,
            systemMetrics = metrics,
        )
    }
}
